cli/lightning-cli.c

```diff
@@ -95,20 +96,139 @@ static size_t human_readable(const char *buffer, const jsmntok_t *t, char term)
 }
 
 static void human_help(const char *buffer, const jsmntok_t *result, bool has_command) {
-	int i;
+	unsigned int i, category;
+	/* The help array contained in the result */
 	const jsmntok_t * help_array = result + 2;
-	/* the first command object */
-	const jsmntok_t * curr = help_array + 1;
-	/* iterate through all commands, printing the name and description */
-	for (i = 0; i < help_array->size; i++) {
+	/* A command object has the following structure :
+	 * {
+	 *     "command": "The command name",
+	 *     "category": "The command category",
+	 *     "description": "The command's description",
+	 *     "verbose": "The command's detailed description"
+	 * }
+	 *
+	 * There are seven categories ("Bitcoin"(0), "Channels"(1), "Network"(2), "Payment"(3),
+	 * "Plugins"(4), "Utilities"(5), "Developer"(6)). We will iterate through all commands,
+	 * adding a pointer to the command token to the corresponding array. We'll then just
+	 * have to iterate through each category's array to printf the commands.
+	 */
+	/* A token array by category */
+	const jsmntok_t **bitcoin_commands = tal_arrz(tmpctx, const jsmntok_t*, 0);
+	const jsmntok_t **channel_commands = tal_arrz(tmpctx, const jsmntok_t*, 0);
+	const jsmntok_t **network_commands = tal_arrz(tmpctx, const jsmntok_t*, 0);
+	const jsmntok_t **payment_commands = tal_arrz(tmpctx, const jsmntok_t*, 0);
+	const jsmntok_t **plugin_commands = tal_arrz(tmpctx, const jsmntok_t*, 0);
+	const jsmntok_t **utility_commands = tal_arrz(tmpctx, const jsmntok_t*, 0);
+	const jsmntok_t **developer_commands = tal_arrz(tmpctx, const jsmntok_t*, 0);
+	/* And a token to iterate through the array */
+	const jsmntok_t * curr;
+
+	json_for_each_arr(i, curr, help_array) {
+		json_to_number(buffer, json_delve(buffer, curr, ".category"), &category);
+		switch(category) {
+		case CMD_BITCOIN:
+			// Add a reference to the token at our array ..
+			tal_arr_expand(&bitcoin_commands, curr);
+			break;
+		case CMD_CHANNELS:
+			tal_arr_expand(&channel_commands, curr);
+			break;
+		case CMD_NETWORK:
+			tal_arr_expand(&network_commands, curr);
+			break;
+		case CMD_PAYMENT:
+			tal_arr_expand(&payment_commands, curr);
+			break;
+		case CMD_PLUGIN:
+			tal_arr_expand(&plugin_commands, curr);
+			break;
+		case CMD_UTILITY:
+			tal_arr_expand(&utility_commands, curr);
+			break;
+		case CMD_DEVELOPER:
+			tal_arr_expand(&developer_commands, curr);
+			break;
+		}
+	}
+
+	/* Loop through each array of *jsmntok_t, sorted by category. Categories are sorted
+	 * alphabetically (apart the developer one which is appended at the end).
+	 */
+
+	printf("=== BITCOIN ===\n\n");
```

Comment:
If you had a string mapping for categories, you could condense this code down a lot by using that as the heading for each sub-genre.

Reply:
Thank you, besides the hint you comment made me understand what @cdecker meant by "array of arrays" :D